lib/metanorma/collection_construct_model.rb

```diff
@@ -0,0 +1,208 @@
+module Metanorma
+  class Collection
+    class << self
+      # @param file [String]
+      # @param collection_model [Hash]
+      # @return [Metanorma::Collection]
+      def parse_model(file, collection_model)
+        if collection_model["bibdata"]
+          bd = Relaton::Cli::YAMLConvertor
+            .convert_single_file(collection_model["bibdata"])
+        end
+
+        mnf  = CollectionManifest.from_yaml collection_model["manifest"]
+        dirs = collection_model["directives"]
+        pref = collection_model["prefatory-content"]
+        fnl  = collection_model["final-content"]
+
+        new(file: file, directives: dirs, bibdata: bd, manifest: mnf,
+            prefatory: pref, final: fnl)
+      end
+
+      # @param Block [Proc]
+      # @note allow user-specific function to run in pre-parse model stage
+      def set_pre_parse_model(&block)
+        @pre_parse_model_proc = block
+      end
+
+      # @param Block [Proc]
+      # @note allow user-specific function to resolve indentifier
+      def set_indentifier_resolver(&block)
+        @indentifier_resolver = block
```

Comment:
I am not convinced that we need to be this flexible, but then again, more flexible is by default better than less flexible, and this lets you deal with documents other than Metanorma XML (although if we're assigning identifiers to documents other than Metanorma XML, we're going to be facing much bigger problems.)

The extraction of document identifiers from Metanorma XML is already implemented in this gem, it is: 

````
    def docid_to_citeas(bib)
      docid = bib.at(ns("./docidentifier[@primary = 'true']")) ||
        bib.at(ns("./docidentifier")) or return
      docid_prefix(docid)
    end
````